### DISCUSSION TITLE
[Security issue] Seeing totpSecretEncoded, totpSecretQrCode and secrets like these exposed in main.js file in keycloak page

### DISCUSSION THREAD

**Discussion by anish-pr** [keycloak/keycloak#31757] [Q&A]
We noticed that login page javascript coming from the keycloak server has a section for totp build/static/js/main.8b4d0521.js file rendered with values -> https://github.com/keycloak/keycloak/blob/5b521173518a80b9ce87ae8fd48bb13e86a657ca/themes/src/main/resources/theme/keycloak.v2/login/login-config-totp.ftl#L22

I replaced the actual values with "value" so as to just depict the issue.

```
totp: {
                    totpSecretEncoded: "value",
                    qrUrl: "#",
                    totpSecretQrCode: "value",
                    manualUrl: "#",
                    totpSecret: "value",
                    otpCredentials: [],
                    supportedApplications: ["FreeOTP", "Google Authenticator"],
                    policy: {
                        algorithm: "HmacSHA1",
                        digits: 6,
                        lookAheadWindow: 1,
                        type: "totp",
                        period: 30
                    }
                }
```

Why does this come with rendered values on the UI ? Is this a security threat?   if so how to avoid this?

**Comment by darius-m** [comment]
It's probably part of the login flow when the user is required to set up their OTP (either via a required action, or a required authentication step when the user has no second authentication factor configured). Have you seen any populated values in the HTML or Javascript files when visiting the site?

Also, make sure to follow proper procedures when reporting possible security issues. The discussions page public and is certainly not the proper channel. Please follow the security policy in these cases - https://github.com/keycloak/keycloak/security.

**Comment by Mizar01** [comment]
Totp secret must be shown to the user to let his device register that secret.
It's not a security issue.

If the totp secret is shown during the login process, but after having configured the Otp already, this can be a security issue.
